src/mlpack/tests/main_tests/preprocess_binarize_test.cpp

```diff
@@ -0,0 +1,110 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <string>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "PreprocessBinarize";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/methods/preprocess/preprocess_binarize_main.cpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+// Utility function to set a parameter and mark it as passed,
+// using copy semantics for lvalues and move semantics for rvalues.
+template<typename T>
+void SetInputParam(const std::string& name, T&& value)
+{
+  CLI::GetParam<typename std::remove_reference<T>::type>(name)
+       = std::forward<T>(value);
+  CLI::SetPassed(name);
+}
+
+struct PreprocessBinarizeTestFixture
+{
+ public:
+  PreprocessBinarizeTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~PreprocessBinarizeTestFixture()
+  {
+    // Clear the settings.
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(PreprocessBinarizeMainTest,
+                         PreprocessBinarizeTestFixture);
+
+/**
+ * Check that input and output have same dimensions.
+ */
+BOOST_AUTO_TEST_CASE(PreprocessBinarizeDimensionTest)
+{
+  // synthetic dataset.
+  arma::mat inputData = arma::randu<arma::mat>(2, 5);
+
+  // Store size of input dataset.
+  size_t input_size  = inputData.n_cols;
+
+  // Input custom data and labels.
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("threshold", (double) 0.5);
+  SetInputParam("dimension", (int) 1);
+
+  mlpackMain();
+
+  // Now check that the output has desired dimensions.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows, 2);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols, input_size);
+}
+
+/**
+ * Check that specified dimension is non-negative.
+ */
+BOOST_AUTO_TEST_CASE(PreprocessBinarizeNegativeDimensionTest)
+{
+  arma::mat inputData = arma::randu<arma::mat>(2, 2);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("threshold", (double) 0.5);
+  SetInputParam("dimension", (int) -2); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check that we can't specify a dimension larger than input.
+ */
+BOOST_AUTO_TEST_CASE(PreprocessBinarizelargerDimensionTest)
+{
+  arma::mat inputData = arma::randu<arma::mat>(2, 2);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("threshold", (double) 0.5);
+  SetInputParam("dimension", (int) 6); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
```

Comment:
There are a couple additional tests I think we could add---let me know what you think.

 - A test to ensure that binarization actually did turn the values into 0 and 1 according to the threshold.
 - A test that ensures all dimensions are binarized when `--dimension` is not specified.

If you don't want to implement these, that's perfectly okay; we can just make a note of it in #1152. :)

Reply:
If I am not wrong, we are already taking care of above two tests through `binarize_test.cpp`. If you want me to implement them here please just let me know.

Reply:
It seems to me like `PreprocessBinarizeNegativeDimensionTest` and `PreprocessBinarizeLargerDimensionTest` both test that the `dimension` parameter is valid, and `PreprocessBinarizeDimensionTest` checks if the output shape of the matrix is correct when only one dimension is specified via `dimension`.  So I think there is no test to ensure that binarization did actually happen, and there is no test to check if all dimensions are binarized.  Let me know if I missed something...

Like I said it's ok if you want to leave it out; these can be added or handled some other time---your call.